File: kluctl/utils/thread_safe_cache.py
import threading


class ThreadSafeCache:
    def __init__(self):
        self.lock = threading.Lock()
        self.cache = {}

    def get(self, key, func):
        with self.lock:
            entry = self.cache.setdefault(key, {
                "done": False,
                "value": None,
                "exception": None
            })

            def do_ret():
                if entry["exception"] is not None:
                    raise entry["exception"]
                return entry["value"]

            if entry["done"]:
                return do_ret()

            try:
                entry["value"] = func()
            except Exception as e:
                entry["exception"] = e
            entry["done"] = True
            return do_ret()

class ThreadSafeMultiCache:
    def __init__(self):
        self.cache = ThreadSafeCache()

    def get(self, cache_key, key, func):
        cache = self.cache.get(cache_key, lambda: ThreadSafeCache())
        return cache.get(key, func)

File: pkg/utils/thread_safe_cache.go
package utils

import "sync"

type ThreadSafeCache[K comparable, V any] struct {
	mutex sync.Mutex
	cache map[K]*cacheEntry[V]
}

type cacheEntry[V any] struct {
	value V
	err   error
}

func (c *ThreadSafeCache[K, V]) Clear() {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	clear(c.cache)
}

func (c *ThreadSafeCache[K, V]) ForEach(cb func(k K, v V)) {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	for k, v := range c.cache {
		if v.err != nil {
			continue
		}
		cb(k, v.value)
	}
}

func (c *ThreadSafeCache[K, V]) Get(key K, f func() (V, error)) (V, error) {
	c.mutex.Lock()
	defer c.mutex.Unlock()

	if c.cache == nil {
		c.cache = make(map[K]*cacheEntry[V])
	}

	e, ok := c.cache[key]
	if ok {
		return e.value, e.err
	}
	e = &cacheEntry[V]{}
	e.value, e.err = f()
	c.cache[key] = e

	return e.value, e.err
}

type ThreadSafeMultiCache[K comparable, V any] struct {
	cache ThreadSafeCache[string, *ThreadSafeCache[K, V]]
}

func (c *ThreadSafeMultiCache[K, V]) Get(cacheKey string, key K, f func() (V, error)) (V, error) {
	cache, _ := c.cache.Get(cacheKey, func() (*ThreadSafeCache[K, V], error) {
		return &ThreadSafeCache[K, V]{}, nil
	})
	return cache.Get(key, f)
}
